ykrt/src/mt.rs

```diff
@@ -139,9 +169,78 @@ impl MT {
         }
     }
 
-    /// `Location`s represent a statemachine: this function transitions to the next state (which
-    /// may be the same as the previous state!). If this results in a compiled trace, it returns
-    /// `Some(pointer_to_trace_function)`.
+    pub fn set_loc_code_ptr(loc: *mut Location, code: *const c_void) {
+        let mt = MT::global();
+        let js = &mut mt.inner.jit_state.lock().unwrap();
+        js.trace_code = code as usize;
+    }
+
+    /// FIXME: This is an interim version of `transition_location()` with (over-)simplified logic.
+    /// The state machines of `Location`s isn't properly hooked in, but this is the proposed
+    /// interface.
+    ///
+    /// This function is called once upon each execution of the control point. The internal state
+    /// of the supplied `Location` (if any) may be mutated, maybe moving the `Location` to a
+    /// different state its state machine.
+    ///
+    /// The returned `usize` signals what action the control point should take next (if any). The
+    /// potential values for this value are as follow:
+    ///
+    ///   +---------------------+---------------+---------------------------+
+    ///   | action   \    bits  |   MSB..=bit2  |      bit1 and bit0        |
+    ///   +---------------------+---------------+---------------------------+
+    ///   | nothing             |       0       | JITACTION_NOP             |
+    ///   | start tracing       |       0       | JITACTION_START_TRACING   |
+    ///   | stop tracing        |       0       | JITACTION_STOP_TRACING    |
+    ///   | execute JITted code |     aligned pointer to JITted code        |
+    ///   +---------------------+-------------------------------------------+
+    ///
+    /// In other words, if the value is neither `JITACTION_NOP`, `JITACTION_START_TRACING`, nor
+    /// `JITACTION_STOP_TRACING`, then the encoded action is to execute JITted code, and the full
+    /// `usize` value encodes a pointer to the executable code.
+    ///
+    /// Values not listed in the table are invalid encodings.
+    ///
+    /// This encoding scheme assumes that the allocator used to obtain heap memory for JITted code
+    /// will align allocations with at least two zero-valued least-significant bits. In other
+    /// words, we assume (and assert) that the `JITACTION_*` values are invalid aligned pointer
+    /// values.
+    ///
+    /// If a null `Location` pointer is passed, then this signals to the JIT that a loop cannot
+    /// start at this point in the interpreter. In this case the function always returns
+    /// `JITACTION_NOP`.
+    pub fn transition_location_simple(loc: *mut Location) -> usize {
+        if loc != ptr::null_mut() {
```

Comment:
It might be clearer (and use less indentation!) to start this function with:

```rust
if loc == ptr::null_mut() {
  return JITACTION_NOP;
}
```

Reply:
Fixed in f9c355c.